src/esp/metadata/managers/SceneAttributesManager.cpp

```diff
@@ -0,0 +1,181 @@
+                        "specification error in scene "
+                     << attribsName << " at idx : " << i << ".";
+      }
+    }
+  } else {
+    LOG(WARNING) << "SceneAttributesManager::setValsFromJSONDoc : No Objects "
+                    "specified for scene "
+                 << attribsName << ", or specification error.";
+  }
+  std::string dfltLighting = "";
+  if (io::jsonIntoVal<std::string>(jsonConfig, "default_lighting",
+                                   dfltLighting)) {
+    // if "default lighting" is specified in scene json set value.
+    attribs->setLightingHandle(dfltLighting);
+  } else {
+    LOG(WARNING)
+        << "SceneAttributesManager::setValsFromJSONDoc : No default_lighting "
+           "specified for scene "
+        << attribsName << ".";
+  }
+
+  std::string navmeshName = "";
+  if (io::jsonIntoVal<std::string>(jsonConfig, "navmesh_instance",
+                                   navmeshName)) {
+    // if "navmesh_instance" is specified in scene json set value.
+    attribs->setNavmeshHandle(navmeshName);
+  } else {
+    LOG(WARNING)
+        << "SceneAttributesManager::setValsFromJSONDoc : No navmesh_instance "
+           "specified for scene "
+        << attribsName << ".";
+  }
+
+  std::string semanticDesc = "";
+  if (io::jsonIntoVal<std::string>(jsonConfig, "semantic_scene_instance",
+                                   semanticDesc)) {
+    // if "semantic scene instance" is specified in scene json set value.
+    attribs->setSemanticSceneHandle(semanticDesc);
+  } else {
+    LOG(WARNING) << "SceneAttributesManager::setValsFromJSONDoc : No "
+                    "semantic_scene_instance specified for scene "
+                 << attribsName << ".";
+  }
+}  // SceneAttributesManager::setValsFromJSONDoc
+
+SceneObjectInstanceAttributes::ptr
+SceneAttributesManager::createInstanceAttributesFromJSON(
+    const io::JsonGenericValue& jCell) {
+  SceneObjectInstanceAttributes::ptr instanceAttrs =
+      SceneObjectInstanceAttributes::create("");
+  // template handle describing stage/object instance
+  io::jsonIntoConstSetter<std::string>(
+      jCell, "template_name",
+      std::bind(&SceneObjectInstanceAttributes::setHandle, instanceAttrs, _1));
+
+  // motion type of object.  Ignored for stage.  TODO : veify is valid motion
+  // type using standard mechanism of static map comparison.
+
+  int motionTypeVal = motionTypeVal =
+      static_cast<int>(physics::MotionType::UNDEFINED);
+  std::string tmpVal = "";
+  if (io::jsonIntoVal<std::string>(jCell, "motion_type", tmpVal)) {
+    // motion type tag was found, perform check - first convert to lowercase
+    std::string strToLookFor = Cr::Utility::String::lowercase(tmpVal);
+    if (SceneObjectInstanceAttributes::MotionTypeNamesMap.count(strToLookFor)) {
+      motionTypeVal = static_cast<int>(
+          SceneObjectInstanceAttributes::MotionTypeNamesMap.at(strToLookFor));
```

Comment:
Minor but very common unneeded double lookup (sorry, another personal check triggerred, heh). If you do the following instead, it'll be just a single lookup:

```suggestion
    auto found = SceneObjectInstanceAttributes::MotionTypeNamesMap.find(strToLookFor);
    if (found != SceneObjectInstanceAttributes::MotionTypeNamesMap.end()) {
      motionTypeVal = static_cast<int>(*found);
```

Alternatively, if all you ever need is converting a string to an enum, to make it easier to use you could have for example a `Containers::Optional<MotionType> motionTypeFromString(const std::string&)` (a free function, even?) that does this lookup and then returns either a valid enum or an optional (and could be doing that just through a cascade of ifs, there's not that many values anyway). Then it'd be just

```cpp
    if (Containers::Optional<MotionType> type = motionTypeFromString(strToLookFor)) {
      motionTypeVal = static_cast<int>(*type);
    } else {
```

Reply:
Thanks mosra! This stuff really helps me. :)